src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,224 @@
+        private static readonly ActivitySource RemotingActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+
+        private static readonly ConcurrentDictionary<string, string> ServiceNameCache = new ConcurrentDictionary<string, string>();
+
+        private readonly RemotingInstrumentationOptions options;
+
+        public TelemetryDynamicSink(RemotingInstrumentationOptions options)
+        {
+            this.options = options;
+        }
+
+        public void ProcessMessageStart(IMessage reqMsg, bool bCliSide, bool bAsync)
+        {
+            try
+            {
+                if (this.options.Filter?.Invoke(reqMsg) == false)
+                {
+                    return;
+                }
+            }
+            catch (Exception e)
+            {
+                RemotingInstrumentationEventSource.Log.MessageFilterException(e);
+                return;
+            }
+
+            try
+            {
+                // Are we executing on client?
+                if (bCliSide)
+                {
+                    // Start new outgoing activity
+                    var act = RemotingActivitySource.StartActivity(ActivityOutName, ActivityKind.Client);
+                    if (act != null)
+                    {
+                        if (act.IsAllDataRequested && reqMsg is IMethodMessage methodMsg)
+                        {
+                            SetStartingActivityAttributes(act, methodMsg);
+                        }
+
+                        var callContext = (LogicalCallContext)reqMsg.Properties["__CallContext"];
+
+                        this.options.Propagator.Inject(new PropagationContext(act.Context, Baggage.Current), callContext, InjectActivityProperties);
+                    }
+                }
+                else
+                {
+                    // We are on server, need to start new or attach to an existing incoming activity.
+
+                    var callContext = (LogicalCallContext)reqMsg.Properties["__CallContext"];
+                    var activityParentContext = this.options.Propagator.Extract(default, callContext, ExtractActivityProperties);
+
+                    Activity ourActivity;
+
+                    // Do we already have an incoming activity?
+                    // Existing activity might be started by an instrumentation layer higher up the
+                    // stack. E.g. if we are using http as a transport for remoting, and we have
+                    // instrumented incoming http request, then that instrumentation happens before
+                    // our "ProcessMessageStart" and we just need to attach to that existing activity.
+                    var parentActivity = Activity.Current;
+                    if (parentActivity != null)
+                    {
+                        ourActivity = RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server);
```

Comment:
I am a bit unsure about this part. If incoming http request is instrumented and creates activity, would it be extracing the same context as the one this instrumentation would have extracted from the callcontext?

Reply:
Thank you! It did sometimes extract incorrect context, in a case when server has aspnet instrumentation enabled but the client didn't enable http instrumentation.

I ended up changing this part quite a bit to cover various cases. I am a bit unsure if `parentActivity.TraceId == callContext.TraceId` check is robust enough if I just want to inherit whatever context the aspnet instrumentation establishes. The "call stack" I am trying to support is:
* Client RemotingOut
  * Client HttpOut
    * Server AspNet
      * Server RemotingIn

I didn't want to just discard the AspNet Activity in this case as it is nice to be able to see all 4 elements of this trace.